### PR TITLE
ui next: Routing for tabbed pages

### PR DESCRIPTION
**Pull request by mrtracy**
Refactor our Routes definition to allow for "tabbed" sub-pages: for example, the
"overview" and "graphs" tab on the Cluster page.
- Adds route definitions for all "tabs" on the Nodes and Cluster page.
- Add "Single Node" page, which does not yet have its tabs.
- Removed all "Title" components; instead, routed components may statically
  define a "Title" function, which the Layout component can find. This is the
  approach suggested by react-router's "breadcrumbs" example.
- Created "ListLink", a general purpose component to create an active
  react-router Link for a List Element. This is more general than the "IconLink"
  previously used by the navigation sidebar.

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/6655)

### CODE REVIEW

**Comment by tamird** [comment]
Reviewed 16 of 16 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[ui/next/app/interfaces/layout.d.ts, line 7 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6655#-KHanHL9DoA-lB5faXH-:-KHanHL9DoA-lB5faXH0:-269303570) ([raw file](https://github.com/cockroachdb/cockroach/blob/d627afd1f473782ae6643e9fb3fb51c961b3b6c5/ui/next/app/interfaces/layout.d.ts#L7)):*
> ```XML
>   title(routeProps: RouteComponentProps<any, any>): React.ReactElement<any>;
> }
> ```

random space

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6655)*

**Comment by maxlang** [comment]
:lgtm:

<details><summary><i>Previously, mrtracy wrote…</i></summary>
> #### ui next: Routing for tabbed pages
> 
> Refactor our Routes definition to allow for "tabbed" sub-pages: for example, the
> 
> "overview" and "graphs" tab on the Cluster page.
> 
> 
> 
> + Adds route definitions for all "tabs" on the Nodes and Cluster page.
> 
> + Add "Single Node" page, which does not yet have its tabs.
> 
> + Removed all "Title" components; instead, routed components may statically
> 
> define a "Title" function, which the Layout component can find. This is the
> 
> approach suggested by react-router's "breadcrumbs" example.
> 
> + Created "ListLink", a general purpose component to create an active
> 
> react-router Link for a List Element. This is more general than the "IconLink"
> 
> previously used by the navigation sidebar.
> 
> 
> </details>


---

Review status: 15 of 16 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[ui/next/app/app.tsx, line 102 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6655#-KHv2LdjTJJkavktcBt1:-KHv2LdjTJJkavktcBt2:1993361871) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d6db2f78777b6bde7799b7e59ee2793138b36d0/ui/next/app/app.tsx#L102)):*
> ```XML
>           <IndexRoute component={ ClusterOverview } />
>           <Route path="events" component={ ClusterEvents } />
>         </Route>
> ```

Just to make sure I'm on the same page. So the route builds the components using the hierarchy and matching routes?

`/cluster/overview ` maps to
```
<Layout>
  <Cluster>
    <ClusterOverview />
  </Cluster>
</Layout>
```

---

*[ui/next/app/containers/nodes.tsx, line 22 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6655#-KHv14uLD_i9U1hsePnP:-KHv14uMOvYrNpwA5ENG:1025773412) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d6db2f78777b6bde7799b7e59ee2793138b36d0/ui/next/app/containers/nodes.tsx#L22)):*
> ```XML
>           <ListLink to="/nodes/graphs">Graphs</ListLink>
>         </ul>
>       </div>
> ```

Does it make sense to pull this into a helper function of some sort?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6655#-:-KHv6JxFwa-JO48cQlpP:1417088721)*

**Comment by mrtracy** [comment]
Review status: 15 of 16 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[ui/next/app/app.tsx, line 102 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6655#-KHv2LdjTJJkavktcBt1:-KHv7Q8MPXNkDgvQGty2:600440412) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d6db2f78777b6bde7799b7e59ee2793138b36d0/ui/next/app/app.tsx#L102)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary>
> Just to make sure I'm on the same page. So the route builds the components using the hierarchy and matching routes?
> 
> `/cluster/overview ` maps to
> ```
> <Layout>
>   <Cluster>
>     <ClusterOverview />
>   </Cluster>
> </Layout>
> ```</details>
Actually, just `/cluster/` maps to that, because `ClusterOverview` renders to the index route.  But yes, that is the structure that is created.

---

*[ui/next/app/containers/nodes.tsx, line 22 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6655#-KHv14uLD_i9U1hsePnP:-KHv73g46DKOK4oWNtbG:-432609305) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d6db2f78777b6bde7799b7e59ee2793138b36d0/ui/next/app/containers/nodes.tsx#L22)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary>
> Does it make sense to pull this into a helper function of some sort?</details>
I think it's simple enough as is, ListLink does the heavy lifting and this is very flexible and readable.  Really, what other information is there here aside from the link target and the text of the links?  There's basically zero boilerplate.

---

*[ui/next/app/interfaces/layout.d.ts, line 7 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6655#-KHanHL9DoA-lB5faXH-:-KHv7bFb8b-FtnwelhKB:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/d627afd1f473782ae6643e9fb3fb51c961b3b6c5/ui/next/app/interfaces/layout.d.ts#L7)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> random space</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6655)*

**Comment by tamird** [comment]
Reviewed 1 of 16 files at r1, 1 of 1 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6655)*

File: ui/next/app/app.tsx [hunk 1]
```diff
@@ -59,10 +58,15 @@ import uiReducer from "./redux/ui";
...
 
 import Layout from "./containers/layout";
-import { ClusterMain, ClusterTitle } from "./containers/cluster";
-import { DatabasesMain, DatabasesTitle } from "./containers/databases";
-import { HelpUsMain, HelpUsTitle } from "./containers/helpus";
-import { NodesMain, NodesTitle } from "./containers/nodes";
+import Cluster from "./containers/cluster";
+import ClusterOverview from "./containers/clusterOverview";
+import ClusterEvents from "./containers/clusterEvents";
+import Databases from "./containers/databases";
```
Comment (tamird) [inline ui/next/app/app.tsx]: this line doesn't compile

Comment (maxlang) [inline ui/next/app/app.tsx]: That's my fault - I have a fix in a PR I'm cleaning up. If you need it to work ASAP just change [this line](https://github.com/cockroachdb/cockroach/blob/master/ui/next/app/containers/databases.tsx#L193) to `export default databasesMainConnected;`

Comment (tamird) [inline ui/next/app/app.tsx]: OK. In your next PR please add `make -C ui/next test` to circle-test.sh (shard 2) to prevent this from recurring

**Pull request merged**